### PR TITLE
Penn Club Rank feature (April Fools!)

### PR DESCRIPTION
**Pull request by julianweng**
Please don't merge until April Fools.

### CODE REVIEW

File: frontend/components/ClubDetails.tsx [hunk 1]
```diff
@@ -44,7 +45,7 @@ const DetailBoolIcon = ({ value, alt }): ReactElement => {
   )
 }
 
-const Details = ({ club }: DetailsProps): ReactElement => {
+const Details = ({ club, showElo }: DetailsProps): ReactElement => {
```
Comment (esinx) [inline frontend/components/ClubDetails.tsx:48]: Same here!

File: frontend/components/ClubDisplay.tsx [hunk 2]
```diff
@@ -58,6 +60,71 @@ const ClubDisplay = ({
...
+                  borderRadius: '6px',
+                  backgroundColor: '#FFFFFF33',
+                  textAlign: 'center',
+                  verticalAlign: 'middle',
+                  lineHeight: '36px',
+                  fontWeight: 'bold',
+                  fontSize: '1.2rem',
+                }}
+              >
+                {`${tier}`}
```
Comment (esinx) [inline frontend/components/ClubDisplay.tsx:114]: I will not allow this

File: frontend/components/ClubRank/RankDropdown.tsx [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+  flex: 1;
+  height: 150px;
+  background-color: #ccc; // Change this color to match your skeleton background
+  animation: ${breatheAnimation} 1.5s ease-in-out infinite;
+`
+
+function RankDropdown(props: RankDropdownProps): ReactElement {
+  const [club1, setClub1] = useState<Club>()
+  const [club2, setClub2] = useState<Club>()
+  const [loading, setLoading] = useState<boolean>(true)
```
Comment (esinx) [inline frontend/components/ClubRank/RankDropdown.tsx:31]: You don't need `<boolean>`

File: frontend/components/ClubRank/RankDropdown.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import styled, { keyframes } from 'styled-components'
+
+import { Club } from '~/types'
+import { doApiRequest } from '~/utils'
+
+import ClubCard from '../ClubCard'
+import { Collapsible } from '../SearchBar'
+
+type RankDropdownProps = {
+  reloadRankings: () => void
```
Comment (esinx) [inline frontend/components/ClubRank/RankDropdown.tsx:11]: Make optional, call it `onReloadRankings?`

File: frontend/components/ClubTableRow.tsx [hunk 5]
```diff
@@ -84,15 +85,18 @@ class ClubTableRow extends React.Component<ClubTableRowProps> {
...
+                  </Name>
                   <div>
                     <TagGroup tags={tags} />
                   </div>
                 </div>
                 <div className="column is-8-desktop is-12-mobile">
                   <Content>
                     <Subtitle>{this.getSubtitle()}</Subtitle>
-                    <ClubDetails club={club} />
+                    <ClubDetails club={club} showElo={showElo} />
```
Comment (esinx) [inline frontend/components/ClubTableRow.tsx:99]: Is this...
image

File: frontend/components/PaginatedClubDisplay.tsx [hunk 6]
```diff
@@ -98,6 +100,7 @@ const PaginatedClubDisplay = ({
           tags={tags}
           display={display}
           onScroll={loadNextPage}
+          ranked={ranked}
```
Comment (esinx) [inline frontend/components/PaginatedClubDisplay.tsx:103]: Is this....
image

File: frontend/pages/clubrank.tsx [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  )
+}
+
+function Rank(props: SplashProps): ReactElement {
+  const [isLoading, setLoading] = useState<boolean>(false)
+  const [searchInput, setSearchInput] = useState<SearchInput>({})
+  const [clubs, setClubs] = useState<PaginatedClubPage>(props.clubs)
+  const currentSearch = useRef<SearchInput>({})
+  useEffect((): void => {
+    const performSearch = async (): Promise<void> => {
```
Comment (esinx) [inline frontend/pages/clubrank.tsx:66]: No type annotation necessary!

File: frontend/pages/clubrank.tsx [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+      if (equal(searchInput, currentSearch.current)) {
+        return
+      }
+      setLoading(true) // Only show loading animation if actively searching, rather than refreshing.
+      await search()
+      setLoading(false)
+    }
+
+    performSearch()
```
Comment (esinx) [inline frontend/pages/clubrank.tsx:75]: consider using IIFC,

```typescript
useEffect(() => {
    (async () => { ... })()
}, [...])
```

**Review by esinx — CHANGES_REQUESTED** [review]
🫡

File: backend/.vscode/settings.json [hunk 9]
```diff
@@ -7,7 +7,7 @@
     "[python]": {
         "editor.formatOnPaste": false,
         "editor.codeActionsOnSave": {
-            "source.fixAll": true
+            "source.fixAll": "explicit"
```
[the remark this answers is not in the document]
Comment (julianweng) [reply backend/.vscode/settings.json:10]: Thank you!

File: frontend/components/ClubCard.tsx [hunk 10]
```diff
@@ -94,50 +94,62 @@ const shorten = (desc: string): string => {
 type ClubCardProps = {
   club: Club
   fullWidth?: boolean
+  clickable?: boolean
 }
 
-const ClubCard = ({ club, fullWidth }: ClubCardProps): ReactElement => {
+const ClubCard = ({
```
[the remark this answers is not in the document]
Comment (julianweng) [reply frontend/components/ClubCard.tsx:100]: Thanks.

File: frontend/components/ClubDetails.tsx (hunk 1 above)
Comment (julianweng) [reply frontend/components/ClubDetails.tsx:48]: Thank you again.

File: frontend/components/ClubDisplay.tsx (hunk 2 above)
Comment (julianweng) [reply frontend/components/ClubDisplay.tsx:114]: image

File: frontend/components/ClubRank/RankDropdown.tsx (hunk 4 above)
Comment (julianweng) [reply frontend/components/ClubRank/RankDropdown.tsx:11]: 🫡

File: frontend/components/ClubTableRow.tsx (hunk 5 above)
Comment (julianweng) [reply frontend/components/ClubTableRow.tsx:99]: 61ZkaCujGrL _AC_UF1000,1000_QL80_

File: frontend/components/PaginatedClubDisplay.tsx (hunk 6 above)
Comment (julianweng) [reply frontend/components/PaginatedClubDisplay.tsx:103]: 61ZkaCujGrL _AC_UF1000,1000_QL80_

File: frontend/pages/clubrank.tsx (hunk 7 above)
Comment (julianweng) [reply frontend/pages/clubrank.tsx:66]: Thanks.

File: frontend/pages/clubrank.tsx (hunk 8 above)
Comment (julianweng) [reply frontend/pages/clubrank.tsx:75]: Thank you for making me a better person than I am today.

File: backend/clubs/management/commands/ban_all_from_clubrank.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+                        self.style.ERROR(
+                            "Please provide a valid argument: ban or unban."
+                        )
+                    )
+                    return
+            Profile.objects.bulk_update(queryset, ["time_since_vote"])
+
+            self.stdout.write(
+                self.style.SUCCESS(
+                    f"Successfully {args[0]}ned everyone from club rank!"
```
[the remark this answers is not in the document]
Comment (julianweng) [reply backend/clubs/management/commands/ban_all_from_clubrank.py]: Used kwargs instead, thanks.

**Comment by julianweng** [comment]
Notes for future years:

1. There are some commits in master (directly following merge commit) with various bug fixes pertaining to this.
2. There's a minor bug where repeatedly ranking clubs would disable the scrolling/pagination feature until page refresh. This is likely because we continue to show the clubdisplay with stale data when refreshing club rank data due to a user ranking a new club, causing issues when a user doesn't wait for the new data before attempting to reach the next paginated page. For future reference, might be good to introduce a ```shouldLoadNextPage``` prop that, upon a change in the clubs prop from parent (signifying new data loaded from an external trigger), allows us to wait on savedNextUrl to be not null before running ```loadNextPage```. 

Added loading indicator while fetching new data as a result of ranking clubs for now, since this change affects the main homepage and reverting master commits could get messy.

**Pull request merged**